core/node/rpc/metrics_interceptor.go

```diff
@@ -3,33 +3,82 @@ package rpc
 import (
 	"context"
 
+	"connectrpc.com/connect"
 	"github.com/prometheus/client_golang/prometheus"
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
-
-	"connectrpc.com/connect"
 )
 
 type streamIdProvider interface {
 	GetStreamId() string
 }
 
-func (s *Service) NewMetricsInterceptor() connect.UnaryInterceptorFunc {
-	interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
-		return func(
-			ctx context.Context,
-			req connect.AnyRequest,
-		) (connect.AnyResponse, error) {
-			proc := req.Spec().Procedure
-			defer prometheus.NewTimer(s.rpcDuration.WithLabelValues(proc)).ObserveDuration()
-
-			r, ok := req.Any().(streamIdProvider)
-			if ok {
-				// this line will enrich the tracing span with the streamId
-				span, _ := tracer.SpanFromContext(ctx)
-				span.SetTag("streamId", r.GetStreamId())
-			}
-			return next(ctx, req)
+type metricsInterceptor struct {
+	rpcDuration       *prometheus.HistogramVec
+	unaryInflight     *prometheus.GaugeVec
+	openClientStreams *prometheus.GaugeVec
+	openServerStreams *prometheus.GaugeVec
+}
+
+func (s *Service) NewMetricsInterceptor() connect.Interceptor {
+	return &metricsInterceptor{
+		rpcDuration:       s.rpcDuration,
+		unaryInflight:     s.metrics.NewGaugeVecEx("grpc_unary_inflight", "gRPC unary calls in flight", "proc"),
+		openClientStreams: s.metrics.NewGaugeVecEx("grpc_open_client_streams", "gRPC open client streams", "proc"),
+		openServerStreams: s.metrics.NewGaugeVecEx("grpc_open_server_streams", "gRPC open server streams", "proc"),
+	}
+}
+
+func (i *metricsInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
+	return func(
+		ctx context.Context,
+		req connect.AnyRequest,
+	) (connect.AnyResponse, error) {
+		var (
+			proc = req.Spec().Procedure
```

Comment:
proc name prepends a "/" which we do not need to pass as part of the label. Maybe remove this prior to creating the label or otherwise sanitize the label before adding it.

```
# HELP river_stream_grpc_unary_inflight gRPC unary calls in flight
# TYPE river_stream_grpc_unary_inflight gauge
river_stream_grpc_unary_inflight{proc="/river.StreamService/Info"} 0
```

Reply:
It would change old behaviour that already measures rpc duration and uses the plain procedure name with the `/` prefix.